src/contracts/MsVault.h

```diff
@@ -0,0 +1,708 @@
+            if (locals.approvals >= 2)
+                locals.releaseApproved = true;
+        }
+
+        if (locals.releaseApproved)
+        {
+            if (locals.vault.balance >= (sint64)input.amount)
+            {
+                qpi.transfer(input.destination, input.amount);
+                locals.vault.balance -= (sint64)input.amount;
+
+                locals.rr_in.vault = locals.vault;
+                resetReleaseRequests(qpi, state, locals.rr_in, locals.rr_out, locals.rr_locals);
+                locals.vault = locals.rr_out.vault;
+
+                state.vaults.set(input.vaultID, locals.vault);
+
+                loggerMsg._type = 4;
+                LOG_INFO(loggerMsg);
+            }
+            else
+            {
+                loggerMsg._type = 5;
+                LOG_INFO(loggerMsg);
+            }
+        }
+        else
+        {
+            state.vaults.set(input.vaultID, locals.vault);
+            loggerMsg._type = 6;
+            LOG_INFO(loggerMsg);
+        }
+    _
+
+    PUBLIC_PROCEDURE_WITH_LOCALS(resetRelease)
+        if (qpi.invocationReward() > RELEASE_RESET_FEE)
+        {
+            qpi.transfer(qpi.invocator(), qpi.invocationReward() - RELEASE_RESET_FEE);
+        }
+        state.totalRevenue += RELEASE_RESET_FEE;
+
+        MSVaultLogger loggerMsg;
+        loggerMsg._contractIndex = CONTRACT_INDEX;
+        loggerMsg._type = 0;
+        loggerMsg.vaultID = input.vaultID;
+        loggerMsg.ownerID = qpi.invocator();
+        loggerMsg.amount = 0;
+        loggerMsg.destination = NULL_ID;
+
+        if (input.vaultID >= MAX_VAULTS)
+        {
+            loggerMsg._type = 1; 
+            LOG_INFO(loggerMsg);
+            return;
+        }
+
+        locals.vault = state.vaults.get(input.vaultID);
+
+        if (!locals.vault.isActive)
+        {
+            loggerMsg._type = 1; 
+            LOG_INFO(loggerMsg);
+            return;
+        }
+
+        locals.io_in.vault = locals.vault;
+        locals.io_in.ownerID = qpi.invocator();
+        isOwnerOfVault(qpi, state, locals.io_in, locals.io_out, locals.io_locals);
+        if (!locals.io_out.result)
+        {
+            loggerMsg._type = 2; 
+            LOG_INFO(loggerMsg);
+            return;
+        }
+
+        locals.fi_in.vault = locals.vault;
+        locals.fi_in.ownerID = qpi.invocator();
+        findOwnerIndexInVault(qpi, state, locals.fi_in, locals.fi_out, locals.fi_locals);
+        locals.ownerIndex = locals.fi_out.index;
+
+        locals.vault.releaseAmounts.set(locals.ownerIndex, 0);
+        locals.vault.releaseDestinations.set(locals.ownerIndex, NULL_ID);
+
+        state.vaults.set(input.vaultID, locals.vault);
+
+        loggerMsg._type = 7;
+        LOG_INFO(loggerMsg);
+    _
+
+    PUBLIC_FUNCTION_WITH_LOCALS(getVaults)
+        output.numberOfVaults = 0;
+        locals.count = 0;
+        for (locals.i = 0; locals.i < 1024; locals.i++)
+        {
+            Vault v = state.vaults.get(locals.i);
```

Comment:
Please assign locals.v, creating objects on the regular function call stack is not permitted.